contracts/staking/fuse/FuseStaking.sol

```diff
@@ -243,54 +243,64 @@ contract FuseStaking is
 			RATIO_BASE + debtToStakers;
-		uint256 daoPartInFuse = earnings - stakersPartInFuse + debtToDAO;
 
-		uint256 totalAmountOfFuseForFuseAcceptingFaucets = _getAmountOfFuseForAllFaucets();
-		uint256 totalFuseToSwap = stakersPartInFuse
-			+ daoPartInFuse - totalAmountOfFuseForFuseAcceptingFaucets;
+		uint256 keeperPartInFuse = (earnings - stakersPartInFuse)
+			- ((earnings - stakersPartInFuse) * (RATIO_BASE - keeperRatio)) / RATIO_BASE;
 
-		uint256[] memory buyResult = _safeBuyGD(
-			totalFuseToSwap,
-			keccak256("totalFuseToSwap")
-		);
+		uint256 daoPartInFuse = (earnings - stakersPartInFuse)
+				- keeperPartInFuse
+				+ debtToDAO
+				- totalAmountOfFuseForFuseAcceptingFaucets;
 
-		debtToStakers = buyResult[2] * PRECISION * stakersPartInFuse / totalFuseToSwap;
-		debtToDAO = buyResult[2] * PRECISION * daoPartInFuse / totalFuseToSwap;
+		uint256 totalFuseToSwap = stakersPartInFuse + daoPartInFuse + keeperPartInFuse;
 
-		uint256 stakersPartInGoodDollar = buyResult[1] * (RATIO_BASE - globalGivebackRatio)
-			/ RATIO_BASE;
+		uint256[] memory buyResult = _buyGD(totalFuseToSwap);
+
+		// avoiding the stack too deep error
+		buyResult[2] = totalFuseToSwap - buyResult[1];
 
-		uint256 daoPartInGoodDollar = buyResult[1] - stakersPartInGoodDollar;
+		debtToStakers = buyResult[2] * PRECISION * stakersPartInFuse / (totalFuseToSwap - keeperPartInFuse);
+		debtToDAO = buyResult[2] * PRECISION * daoPartInFuse / (totalFuseToSwap - keeperPartInFuse);
 
-		uint256 keeperPartInGoodDollar = daoPartInGoodDollar
-			- (daoPartInGoodDollar * (RATIO_BASE - keeperRatio)) / RATIO_BASE;
+		uint256 stakersPartInGoodDollar = buyResult[1] * PRECISION * stakersPartInFuse / totalFuseToSwap;
+		uint256 daoPartInGoodDollar = buyResult[1] * PRECISION * daoPartInFuse / totalFuseToSwap;
+		uint256 keeperPartInGoodDollar = buyResult[1] - (stakersPartInGoodDollar + daoPartInGoodDollar);
 
 		daoPartInGoodDollar -= keeperPartInGoodDollar;
 
 		uint256 communityPoolPartInGoodDollar = daoPartInGoodDollar
 			- (daoPartInGoodDollar * (RATIO_BASE - communityPoolRatio))
 			/ RATIO_BASE;
 
-		daoPartInGoodDollar -= communityPoolPartInGoodDollar;
+		require(
+			goodDollar.transfer(msg.sender, keeperPartInGoodDollar),
+			"keeperPartTransferFailed"
+		);
+
+		communityPoolBalance += communityPoolPartInGoodDollar;
+
+		uint256 ubiPartInGoodDollar = daoPartInGoodDollar - communityPoolPartInGoodDollar;
 
 		require(
-			goodDollar.transfer(address(ubiScheme), keeperPartInGoodDollar),
+			goodDollar.transfer(address(ubiScheme), ubiPartInGoodDollar),
 			"ubiPartTransferFailed"
 		);
-		communityPoolBalance += communityPoolPartInGoodDollar;
 
-		_distributeGDToFaucets(daoPartInGoodDollar);
+
+		_distributeGDToFaucets(daoPartInGoodDollar - ubiPartInGoodDollar);
 		_distributeFuseToFaucets(totalAmountOfFuseForFuseAcceptingFaucets);
 
 		_updateGlobalGivebackRatio();
 		_notifyRewardAmount(stakersPartInGoodDollar);
```

Comment:
updates should happen before transfers to avoid re-entrancy issues

Reply:
done